content/posts/redis-modules-into-arm-land-part1.md

```diff
@@ -0,0 +1,276 @@
+Finally, [**RedisAI**](https://github.com/RedisAI/RedisAI). It's a C library built with CMake, and it includes modular "engines" that allow abstraction and encapsulation of AI libraries like [TensorFlow](https://github.com/tensorflow/tensorflow), [PyTorch](https://github.com/pytorch/pytorch), and [ONNXRuntime](https://github.com/microsoft/onnxruntime), in their C library form — most users typically use them in Python, with PyTorch and ONNXRuntime not officially supporting ARM.
+So the build requirements, as it seems, have deteriorated quickly. It went from building an innocent C library to compiling massive source bases with convoluted build systems.
+
+In the following sections, we will fit each component with its proper build method.
+
+## Building for ARM
+
+At this point, we'll pause and take stock of what is required to build software for ARM.
+The obvious way is to use an ARM-based device, like Raspberry Pi. Once set up, you'll be able to build and test in the most natural manner. 
+
+The testing medium is important: even though you can build almost every ARM software without a physical ARM device, there is no way to reliably test the outcome without such a device, especially with non-standard devices. Therefore, while a virtualized/emulated/containerized test may be useful, you should always test your software on its designated target device.
+
+A Raspberry Pi 4 with 4GB RAM (and a 1Gbps NIC, no less importantly) and a fast microSD card looks promising.
+
+Regarding the OS selection on ARM: the offerings are limited, and the rule of thumb is to go for the latest release of Raspbian (which is customized Debian distribution for RPi), Ubuntu, or Fedora, the latter two offering easy-to-install ARM systems. Do not worry about immaturity: it was proven time and again that newer systems work better and old ones can't keep up.
+
+Installing an OS brings us to the first dilemma: 
+If we install Raspbian, we get a 32-bit OS (i.e., arm32v7 platform).
+If we choose 64-bit Ubuntu, we get an arm64v8 platform (we discuss ARM platforms in detail later). 
+If we intend to support both platforms (as we do with RedisEdge), we can either get two SD cards and install each OS on its own card while taking turns on the device, or get two RPi devices (which is not terribly expensive). I recommend the latter.
+
+And now, for the principal principle of OS selection: our goal is to have a stable system with the newest Docker version. Any OS that will satisfy these requirements will do.
+That's right: we're going to use Docker to fetch the OS we really wish to build for, using the underlying OS as infrastructure. This will also help keep our build experiments properly isolated.
+
+If you have a RPi device at your disposal, you can now proceed with making it functional. In the next post, as we will present build methods that do not require a physical ARM machine.

```

Comment:
Change to "If you have a RPi device at your disposal, you can now proceed with making it functional. In the next post, we will present build methods that do not require a physical ARM machine."